Miguel/ava 363 proposals on nomad bridge security

MiguelDD1 opened the pull request:
# Security
- `do_dispatch` has been re-written to avoid modification on storage  and failure, as it is not transactional.
- `do_update` needs to be `transactional` to avoid inconsistency between storages if the internal loop fails.
- `update` uses a `max_index_witness` to calculate the proper weight.

# Benchmarks
- Worst case on `nomad-home::update` is when `Tree` has `TREE_DEPTH (=32)` as depth. We use `init_tree` fn to generate that case on the `update` benchmark.
- As `Updater` contains NON compatible WASM code, we use pre-images on benchmarks: `fn expected_signed_update()` and `fn expected_longest_tree_signed_update()`. There are UTs to ensure their values are correct.

# Other changes
- [Reduce fees] `da-bridge::try_dispatch_data_root` uses the block number & the data root instead the whole header.
- `Home` crate has been renamed to `Nomad-Home`.

luketchang committed:
add nomad primitives

luketchang committed:
patch dependencies to fix build issues

luketchang committed:
port nomad pallets, build failing

luketchang committed:
add node and runtime changes

luketchang committed:
add subxt changes

luketchang committed:
fix compact block header and add specs to node for genesis info

luketchang committed:
add new script for submitting data and header

luketchang committed:
rename enqueue_data_root to dispatch_data_root

luketchang committed:
remove tiny keccak from merkle

luketchang committed:
add tiny keccak back to merkle and use to avoid extra allocation in left/right hash

luketchang committed:
rebase on primitive changes and change da-bridge message and TypedMessage to use BoundedVec

luketchang committed:
fix failing home test after max message len change

luketchang committed:
rebase on merged primitives

luketchang committed:
remove unnecessary methods and add da-bridge docs

luketchang committed:
update metadata and have home tree put default

luketchang committed:
make da-bridge message type struct-based not vector-based

luketchang committed:
remove panic in da bridge message

luketchang committed:
remove finalized block hash mapping

migroslinx committed:
Merge remote-tracking branch 'matic/develop' into luke/nomad-dev-pallets

migroslinx committed:
Safety review

luketchang commented on pallets/nomad/da-bridge/src/lib.rs:
```diff
@@ -135,18 +138,11 @@ pub mod pallet {
-			let stored_hash = frame_system::Pallet::<T>::block_hash(number);
-
-			// Ensure header's hash matches that in the block number to hash
-			// mapping
-			let hash = header.hash();
+			let hash = frame_system::Pallet::<T>::block_hash(number);
 			ensure!(
-				stored_hash == hash,
+				hash != Default::default(),
```
Removing the header hash check is not safe. This would allow anyone to pass an arbitrary block header with a fake block number. 

Example: I make up a fake header with block number 0 that doesn't actually exist in the chain. Pallet would only check block number has some mapping but doesn't check that the hash matches to header's hash. Checks pass even though header is fake.

MiguelDD1 replied:
I see.
The thing is that `Header` is really a _heavy_ structure, i.e. it contains the Kate commitment, and I would also like to _remove the hash calculation_ inside the runtime.
What do you thing about sending only the 3 fields used internally? I mean:
- header hash,
- block number, and
- data root.

Something like:
```Rust
pub fn try_dispatch_data_root(
   origin: OriginFor<T>,
   destination_domain: u32,
   #[pallet::compact] destination_domain: u32,
   recipient_address: H256,
   header_hash: T::Hash,
   #[pallet::compact] block_number: T::BlockNumber,
   data_root: H256 ) -> DispatchResult
```

luketchang replied:
Sure that should work. We can get stored_hash from the `block_hash` mapping and check that the stored hash == header_hash

luketchang commented on pallets/nomad/home/src/lib.rs:
```diff
@@ -299,8 +308,6 @@ pub mod pallet {
 			signed_update: &SignedUpdate,
 		) -> Result<bool, DispatchError> {
-			Self::ensure_not_failed()?;
-
```
We want to keep this, as `fn improper_update` is still an exposed method separate from `fn update` so not guaranteed `improper_update` always come from an `update` call

luketchang commented on pallets/nomad/home/src/lib.rs:
```diff
@@ -162,9 +163,13 @@ pub mod pallet {
+		pub fn update(
+			origin: OriginFor<T>,
+			signed_update: SignedUpdate,
+			#[pallet::compact] max_index: u32,
```
Will need to check that we are able to format dynamic transactions that scale encode `compact`

MiguelDD1 replied:
It should be because `Header::Number` is compact too.

migroslinx committed:
Add benchmarks to `Nomad-home`

migroslinx committed:
Merge branch 'luke/nomad-dev-pallets' of github.com-miguelDD1:nomad-xyz/avail into luke/nomad-dev-pallets

migroslinx committed:
Merge branch 'luke/nomad-dev-pallets' into miguel/ava-363-proposals-on-nomad-bridge-security

luketchang commented on pallets/nomad/da-bridge/src/lib.rs:
```diff
@@ -135,18 +138,11 @@ pub mod pallet {
 			ensure!(
-				stored_hash == hash,
+				hash != Default::default(),
 				Error::<T>::HashOfBlockNotMatchBlockNumber,
```
Security: We still want to check that the header provided has a hash that matches the one retrieved from the block number --> block hash mapping. Otherwise, a malicious actor could provide a make up a header with a random hash that was never in the history. Would put back original checks

luketchang commented on pallets/nomad/home/src/benchmarking.rs:
```diff
@@ -1,20 +1,149 @@
+		let signed_update = expected_signed_update();
+
+	}: _(origin, signed_update)
+	verify {
+		// TODO
+	}
```
Flagging TODOs in case they were missed

migroslinx committed:
Code review changes

  - New Benchmarks for `nomad/da-bridge`
  - Add header hash to `try_dispatch_data_root`.
  - Add assertions to `verify` section in benchmarks.

luketchang commented:
Also if you could update the subxt scripts to use the new metadata, that'd be great. Each time signature of method changes, we need to update this

migroslinx committed:
Update `avail-subxt` examples

luketchang commented on pallets/nomad/da-bridge/src/lib.rs:
```diff
@@ -102,11 +108,9 @@ pub mod pallet {
 			recipient_address: H256,
-			header: T::Header,
+			block_number: T::BlockNumber,
+			data_root: H256,
```
data_root is completely unlinked to header hash and block number. This means a bad actor could just make up some arbitrary data root as long as it found a header hash and block number that work

MiguelDD1 replied:
I cannot find a simple way to access to the header info from inside the runtime, so I've reverted this change. The whole header must be send.

MiguelDD1 committed:
Full header is needed

MiguelDD1 committed:
Fmt fix
